src/librustc_mir/dataflow/impls/borrows.rs

```diff
@@ -236,9 +285,65 @@ impl<'a, 'gcx, 'tcx> BitDenotation for Borrows<'a, 'gcx, 'tcx> {
         self.kill_loans_out_of_scope_at_location(sets, location);
     }
 
-    fn terminator_effect(&self,
-                         sets: &mut BlockSets<BorrowIndex>,
-                         location: Location) {
+    fn statement_effect_on_activations(&self,
+                                       sets: &mut BlockSets<ResActIndex>,
+                                       location: Location) {
+        let block = &self.mir.basic_blocks().get(location.block).unwrap_or_else(|| {
+            panic!("could not find block at location {:?}", location);
+        });
+        let stmt = block.statements.get(location.statement_index).unwrap_or_else(|| {
+            panic!("could not find statement at location {:?}");
+        });
+
+        // INVARIANT: At this point, sets.on_entry should correctly
+        // reflect the reservations as we enter the statement.
+
+        // Now compute the effect of the statement on the activations
+        // themselves in the ActiveBorrows state.
+
+        // First: Any uses of reserved Lvalues in the statement are now activated.
+        {
+            let mut find = FindPlaceUses { sets, assigned_map: &self.assigned_map };
+            find.visit_statement(location.block, stmt, location);
+        }
+
+        // Second: EndRegion kills any active borrows.
+        match stmt.kind {
+            mir::StatementKind::EndRegion(region_scope) => {
```

Comment:
Man, I think I've never recommended *against* an exhaustive match before, but this seems like a classic case for `if let`.

Reply:
(Turns out, no; we need to handle other cases here besides `EndRegion`...)